### PR TITLE
feat: add `sendRawTransactionWithDetailedOutput` API

### PR DESCRIPTION
**Pull request by ischasny**
Add a new `zks_sendRawTransactionWithDetailedOutput` API that instantly returns storage logs and events (like if the transaction would have already been applied) alongside the transaction hash. The API is behaviourally analogous to `eth_sendRawTransaction` but with some extra data returned from it.

This feature will allow consumer apps to apply "optimistic" events in their applications instantly without having to wait for zkSync block confirmation time. 

It’s expected that optimistic logs of two uncommitted transaction that modify the same state will not have causal relationships between each other.

### CODE REVIEW

File: core/lib/types/src/api/mod.rs [hunk 1]
```diff
@@ -736,3 +737,52 @@ pub struct Proof {
...
+    #[serde(rename = "storageLogs")]
+    pub storage_logs: Vec<ApiStorageLogQuery>,
+    pub events: Vec<ApiVmEvent>,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ApiVmEvent {
+    #[serde(rename = "l1BatchNumber")]
+    pub l1_batch_number: L1BatchNumber,
```
[the remark this answers is not in the document]
Comment (ischasny) [reply core/lib/types/src/api/mod.rs:756]: Good point - fixed

File: core/lib/zksync_core/src/api_server/execution_sandbox/testonly.rs [hunk 2]
```diff
@@ -47,19 +47,42 @@ impl MockTransactionExecutor {
...
+    where
+        F: Fn(&Transaction, &BlockArgs) -> ExecutionResult + 'static + Send + Sync,
+    {
+        Box::new(
+            move |tx: &Transaction, ba: &BlockArgs| -> VmExecutionResultAndLogs {
+                VmExecutionResultAndLogs {
+                    result: responses(tx, ba),
+                    logs: Default::default(),
+                    statistics: Default::default(),
+                    refunds: Default::default(),
```
[the remark this answers is not in the document]
Comment (ischasny) [reply core/lib/zksync_core/src/api_server/execution_sandbox/testonly.rs:70]: Fixed

File: core/lib/zksync_core/src/api_server/web3/tests/vm.rs [hunk 3]
```diff
@@ -254,6 +262,104 @@ async fn send_raw_transaction_after_snapshot_recovery() {
     .await;
 }
 
+#[derive(Debug)]
+struct SendTransactionOptimisticTest {}
```
Comment (montekki) [inline core/lib/zksync_core/src/api_server/web3/tests/vm.rs:266]:
```suggestion
struct SendTransactionOptimisticTest;
```

Empty structures do not need curlys.

File: core/lib/zksync_core/src/api_server/web3/tests/vm.rs [hunk 4]
```diff
@@ -254,6 +262,104 @@ async fn send_raw_transaction_after_snapshot_recovery() {
...
+#[async_trait]
+impl HttpTest for SendTransactionOptimisticTest {
+    fn transaction_executor(&self) -> MockTransactionExecutor {
+        let mut tx_executor = MockTransactionExecutor::default();
+        let tx_bytes_and_hash = SendRawTransactionTest::transaction_bytes_and_hash();
+        let vm_execution_logs = VmExecutionLogs {
+            storage_logs: self.storage_logs(),
+            events: self.vm_events(),
+            user_l2_to_l1_logs: Default::default(),
+            system_l2_to_l1_logs: Default::default(),
```
Comment (montekki) [inline core/lib/zksync_core/src/api_server/web3/tests/vm.rs:306]: again `..default::Default()`

File: core/lib/zksync_core/src/api_server/web3/tests/vm.rs [hunk 5]
```diff
@@ -254,6 +262,104 @@ async fn send_raw_transaction_after_snapshot_recovery() {
...
+
+        tx_executor.set_tx_responses_with_logs(move |tx, block_args| {
+            assert_eq!(tx.hash(), tx_bytes_and_hash.1);
+            assert_eq!(block_args.resolved_block_number(), L2BlockNumber(1));
+
+            VmExecutionResultAndLogs {
+                result: ExecutionResult::Success { output: vec![] },
+                logs: vm_execution_logs.clone(),
+                statistics: Default::default(),
+                refunds: Default::default(),
```
Comment (montekki) [inline core/lib/zksync_core/src/api_server/web3/tests/vm.rs:318]: Same here

File: core/lib/zksync_core/src/api_server/web3/tests/vm.rs (hunk 4 above)
Comment (ischasny) [reply core/lib/zksync_core/src/api_server/web3/tests/vm.rs:306]: Leaving that as is as discussed

File: core/lib/zksync_core/src/api_server/web3/tests/vm.rs (hunk 5 above)
Comment (ischasny) [reply core/lib/zksync_core/src/api_server/web3/tests/vm.rs:318]: Leaving that as is as discussed

**Review by montekki — APPROVED** [review]

File: core/lib/types/src/api/mod.rs [hunk 6]
```diff
@@ -736,3 +737,38 @@ pub struct Proof {
...
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ApiVmEvent {
+    pub l1_batch_number: L1BatchNumber,
+    pub tx_batch_index: u32,
+    pub address: Address,
+    pub indexed_topics: Vec<H256>,
+    pub value: String,
```
Comment (RomanBrodetski) [inline core/lib/types/src/api/mod.rs:756]: Are you sure we need a new `struct` for this? We already return logs in API, please check out `Log` struct. 

I understand it has a larger set of fields, but IMO consistency is very important here. These logs will be processed by Lens the same way as they process them in transaction receipt, so their code would need to understand both notations (e.g. `indexed_topics` vs `topics`, `transaction_index` vs `tx_batch_index` etc. 

Additionally, I think `struct Log` follows Ethereum specification.

Comment (ischasny) [reply core/lib/types/src/api/mod.rs:756]: makes sense - will use `Log` instead. We will have to use a custom struct for vm events though, yes?

File: core/lib/types/src/api/mod.rs [hunk 7]
```diff
@@ -736,3 +737,38 @@ pub struct Proof {
...
+    pub shard_id: u8,
+    pub address: Address,
+    pub key: U256,
+    pub read_value: U256,
+    pub written_value: U256,
+    pub rw_flag: bool,
+    pub rollback: bool,
+    pub is_service: bool,
+    pub log_type: StorageLogQueryType,
+}
```
Comment (RomanBrodetski) [inline core/lib/types/src/api/mod.rs:774]: Most of these fields are strictly internal, with some being only useful for proof generation. What lens is really interested in is only the updates to storage, which is `(address, key) => new_value` mapping.

File: core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/namespaces/zks.rs [hunk 8]
```diff
@@ -174,4 +175,30 @@ impl ZksNamespaceServer for ZksNamespace {
         self.get_base_token_l1_address_impl()
             .map_err(|err| self.current_method().map_err(err))
     }
+
+    async fn send_raw_transaction_optimistic(
```
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/namespaces/zks.rs:179]: I'm not sure about the naming here. I know that Lens refer to this API as `Optimistic`, but in reality it only refers to the fact that we return VM logs/events that result from the VM execution. There is nothing optimistic about the tx execution itself - the `send_raw_transaction` is executed the same way as normal `send_raw_transaction`.

So I would call it something like `send_raw_transaction_with_detailed_output` or something like `send_raw_transaction_return_execution_logs`.

File: core/lib/types/src/api/mod.rs [hunk 9]
```diff
@@ -740,35 +739,16 @@ pub struct Proof {
...
-    pub tx_batch_index: u32,
-    pub address: Address,
-    pub indexed_topics: Vec<H256>,
-    pub value: String,
+    pub events: Vec<Log>,
 }
 
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
 pub struct ApiStorageLogQuery {
```
Comment (RomanBrodetski) [inline core/lib/types/src/api/mod.rs:750]: let's rename to `ApiStorageLog` (`Query` comes from crypto code and I think it's out of place here)

File: core/lib/types/src/event/mod.rs [hunk 10]
```diff
@@ -42,14 +43,21 @@ impl VmEvent {
...
+        Log {
             address: vm_event.address,
-            indexed_topics: vm_event.indexed_topics.clone(),
-            value: hex::encode(&vm_event.value),
+            topics: vm_event.indexed_topics.clone(),
+            data: Bytes::from(vm_event.value.clone()),
+            block_hash: None,
+            block_number: None,
+            l1_batch_number: Some(U64::from(vm_event.location.0 .0)),
+            transaction_hash: None,
```
Comment (RomanBrodetski) [inline core/lib/types/src/event/mod.rs:55]: nit: do we have any meaningful `transaction_hash` that we can put here? In other words, is there a transaction hash already computed by this point? If not, feel free to keep as is

File: core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/namespaces/zks.rs (hunk 8 above)
Comment (ischasny) [reply core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/namespaces/zks.rs]: Changed to `send_raw_transaction_with_detailed_output`

File: core/lib/types/src/api/mod.rs (hunk 7 above)
Comment (ischasny) [reply core/lib/types/src/api/mod.rs:754]: Fixed - filtering out all read only logs as discussed

File: core/lib/types/src/api/mod.rs (hunk 6 above)
Comment (ischasny) [reply core/lib/types/src/api/mod.rs]: Fixed, as discussed

**Review by RomanBrodetski — DISMISSED** [review]

File: core/lib/types/src/api/mod.rs (hunk 9 above)
Comment (ischasny) [reply core/lib/types/src/api/mod.rs:750]: Renamed

**Review by RomanBrodetski — APPROVED** [review]

**Review by RomanBrodetski — APPROVED** [review]

**Pull request merged**